### PR TITLE
Alarm trigger

### PR DESCRIPTION
**Pull request by fdegier**
Support for triggered alarm. I have tested it an come to the following observations. I have setup a spare Jablotron central, it has no modules connected and the lid is off, therefore it is continuous `tamper` alert. I have added both `alarm` and `tamper` as states for a triggered alarm.

The good:
- It gives a notification if the alarm is triggered
- It shows the alarm is triggered
- Its a trigger per section

The bad:
- Once homebridge starts up, the state does not go to `triggered` after 60 seconds it does trigger the triggered alarm
- After initial start up, it randomly notifies that the alarm is triggered (note this alarm is always triggered)

My thoughts and ask for help:
- I think the initial startup skips the triggering part, however I can not find it in the code
- The background update, it should run every 60 seconds but it does not seem to trigger ever 60 seconds.

@misncz @grzegorz914 can you check the code and chime in on the issues described above?

I can share the Jablotron service so you can play around with it.

### CODE REVIEW

**Comment by fdegier** [comment]
IMG_2346
IMG_2347

**Comment by fdegier** [comment]
Closes #40

**Comment by fdegier** [comment]
@grzegorz914 and @misncz can you please check my PR?

**Review by grzegorz914 — APPROVED** [review]

File: lib/jablotron.js [hunk 1]
```diff
@@ -74,8 +73,11 @@ Jablotron.prototype = {
             if (sessionId && sessionId != null) {
                 self.client.doAuthenticatedRequest('/dataUpdate.json', payload, sessionId, function (response) {
                     let responseData = self.parseResponseData(response);
-                    if (responseData != null) {
-                        let segments = responseData[0]['data']['segments'];
+                    if (responseData["data"] != null) {
+                        if (responseData["service_events_type"] == "tamper" || responseData["service_events_type"] == "alarm") {
```
Comment (misncz) [inline lib/jablotron.js:77]: Be carefull with these attributes - they could be null for thermometers. Even though thermometers use the same structure majority of attributes returned is null

File: lib/jablotron.js [hunk 2]
```diff
@@ -101,15 +103,25 @@ Jablotron.prototype = {
             if (sessionId && sessionId != null) {
                 self.client.doAuthenticatedRequest('/dataUpdate.json', payload, sessionId, function (response) {
                     let responseData = self.parseResponseData(response);
-                    if (responseData != null) {
-                        let segments = responseData[0]['data']['segments'];
+                    if (responseData["data"] != null) {
+                        let alarm_service_event = [];
+                        if (responseData["service_events_type"] == "tamper" || responseData["service_events_type"] == "alarm") {
```
Comment (misncz) [inline lib/jablotron.js:108]: Be carefull with these attributes - they could be null for thermometers. Even though thermometers use the same structure majority of attributes returned is null

File: lib/jablotron.js (hunk 1 above)
Comment (fdegier) [reply lib/jablotron.js:77]: I assume you mean `if (responseData["service_events_type"]` ? This is on section level, I don't think it should influence the operation. 

Do you still have the thermometer for testing?

Comment (fdegier) [reply lib/jablotron.js:77]: Ok, I will update the other PR for merging, can you tomorrow check if it works?

Comment (misncz) [reply lib/jablotron.js:77]: Yep, no problem

**Comment by fdegier** [comment]
@misncz I have resolved the merge conflicts. Since there is now a third value in the callback which for a triggered alarm is not available, I have just added an empty string. Is this OK? Do you mind checking the merge result, especially in `jablotron.js`?

**Comment by misncz** [comment]
> 

Will have a look in the morning. Instead of empty string pass null ( For thermomemters it’s an array of information objects)

File: lib/jablotron.js [hunk 3]
```diff
@@ -104,10 +106,17 @@ Jablotron.prototype = {
...
+                            alarm_service_event =  responseData["service_events"];
+                        }
+                        for (let i = 0; i < responseData["data"].length; i++) {
+                            let segments = responseData["data"][0]['data']['segments'];
                             segments.forEach(function (segment) {
+                                if (alarm_service_event.join().includes(segment['segment_name'])) {
+                                    callback(segment['segment_key'], "triggered", "");
```
Comment (fdegier) [inline lib/jablotron.js:118]:
```suggestion
                                    callback(segment['segment_key'], "triggered", null);
```

**Comment by misncz** [comment]
jablotron.js changes broke thermometer and PGM (sirene or switch). Need to have a look

**Comment by misncz** [comment]
Change to parseResponseData method broke getAccessoryStates method that now loop only through first data type (ie sections) and ignores PGM & thermometers - it's something I'm able to fix
Also getAccessorState looks suspicious to me as it calls callback with 'triggered' but doesn't check accessory key. Also call to callback needs 2 params callback('triggered', null)

**Comment by misncz** [comment]
@fdegier I pushed a fix getAccessoryStates and tried to modify getAccessoryState the same way as getAccessoryStates. Now it should be fine. Please test alarm triggering now

**Comment by fdegier** [comment]
Thank you, good improvements!

I've tested it but now it gives an error that I fixed earlier: 

```
[2/27/2022, 10:58:41 AM] [homebridge-jablotron] This plugin generated a warning from the characteristic 'Security System Target State': characteristic was supplied illegal value: number 4 exceeded maximum of 3. See https://git.io/JtMGR for more info.
```

This happens because it tries to set the target state to `triggered` which isn't a target state, only a current state. I'm investigating the exact cause now.

**Comment by fdegier** [comment]
This part is the cause for the warning: 

```
if (fullReport && state != this.getTriggeredStateMapping()) {
                     this.updateServiceCharacteristic(Characteristic.SecuritySystemTargetState, state);
                 }
```

So state is probably incorrect.

**Comment by MartinKristof** [comment]
Could be merged please?

**Comment by MikeFParkin** [comment]
Does the alarm triggering still not work? ☹️ Will this issue ever be fixed/merged?

**Comment by fdegier** [comment]
It will, eventually. It was ready to be merged but there is now a conflict that breaks the functionality. Besides that we need to upgrade the API version which has a higher priority and will break more things.

**Comment by MikeFParkin** [comment]
> It will, eventually. It was ready to be merged but there is now a conflict that breaks the functionality. Besides that we need to upgrade the API version which has a higher priority and will break more things.

No problem, thanks for the update - and for all the hard work you, and the team, put into making this plugin a reality!

**Comment by fdegier** [comment]
Closing this PR as the functionality will be picked up in a new PR that is on par with the new API version.

**Pull request closed without merging**